# Data Aggregation and Grouping
## Using World Flags Data

<img src='flags.JPG'>

A critical task for data analysis is often aggregating or transforming groups of data. After preparing your data, you may need to compute group statistics or possible pivot tables for reporting or visualization purposes. Pandas `groupby` is a flexible way to perform these aggregations and summarize datasets.

For this module, we will be working with data that contains details of various nations and their flags. It was originally collected from the 'Collins Gen Guide to Flags' from Collins Publishers in 1986. Note that this data is out-of-date. For instance, it still includes 'USSR' as a country.
      
Here is some basic information about the dataset:

- There are 194 instances (aka rows).
- There are 30 attributes in total (aka columns).
- 10 attributes are numeric-valued.  The remainder are either Boolean or nominal-valued.
- There are no missing values.

**Attribute Information**

1. name: Name of the country concerned
2. landmass: 1=N.America, 2=S.America, 3=Europe, 4=Africa, 5=Asia, 6=Oceania
3. zone: Geographic quadrant, based on Greenwich and the Equator (1=NE, 2=SE, 3=SW, 4=NW)
4. area: in thousands of square km
5. population: in round millions
6. language: 1=English, 2=Spanish, 3=French, 4=German, 5=Slavic, 6=Other Indo-European, 7=Chinese, 8=Arabic,            9=Japanese/Turkish/Finnish/Magyar, 10=Others
7. religion: 0=Catholic, 1=Other Christian, 2=Muslim, 3=Buddhist, 4=Hindu, 5=Ethnic, 6=Marxist, 7=Others
8. bars: Number of vertical bars in the flag
9. stripes: Number of horizontal stripes in the flag
10. colors: Number of different colors in the flag
11. red: 0 if red absent, 1 if red present in the flag
12. green: same for green
13. blue: same for blue
14. gold: same for gold (also yellow)
15. white: same for white
16. black: same for black
17. orange: same for orange (also brown)
18. mainhue: predominant colour in the flag (tie-breaks decided by taking the topmost hue, if that fails then the most central hue, and if that fails the leftmost hue)
19. circles: Number of circles in the flag
20. crosses: Number of (upright) crosses
21. saltires: Number of diagonal crosses
22. quarters: Number of quartered sections
23. sunstars: Number of sun or star symbols
24. crescent: 1 if a crescent moon symbol present, else 0
25. triangle: 1 if any triangles present, 0 otherwise
26. icon: 1 if an inanimate image present (e.g., a boat), otherwise 0
27. animate: 1 if an animate image (e.g., an eagle, a tree, a human hand) present, 0 otherwise
28. text: 1 if any letters or writing on the flag (e.g., a motto or slogan), 0 otherwise
29. topleft: color in the top-left corner (moving right to decide tie-breaks)
30. botright: color in the bottom-left corner (moving left to decide tie-breaks)

## Initial Imports

In [2]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', 50)

## Initial Exploration of Flags Dataset

In [4]:
# create list of column names
columns = ['name','landmass','zone','area','population','language','religion','num_bars','num_stripes','num_colors',
           'red','green','blue','gold','white','black','orange','mainhue','num_circles','num_crosses','num_saltires',
           'num_quarters','num_sunstars','crescent','triangle','icon','animate','text','topleft_color','botright_color']

# import data and show first five rows
flags = pd.read_csv('flag.data', names=columns)
flags.head()

,name,landmass,zone,area,population,language,religion,num_bars,num_stripes,num_colors,red,green,blue,gold,white,black,orange,mainhue,num_circles,num_crosses,num_saltires,num_quarters,num_sunstars,crescent,triangle,icon,animate,text,topleft_color,botright_color
0,Afghanistan,5,1,648,16,10,2,0,3,5,1,1,0,1,1,1,0,green,0,0,0,0,1,0,0,1,0,0,black,green
1,Albania,3,1,29,3,6,6,0,0,3,1,0,0,1,0,1,0,red,0,0,0,0,1,0,0,0,1,0,red,red
2,Algeria,4,1,2388,20,8,2,2,0,3,1,1,0,0,1,0,0,green,0,0,0,0,1,1,0,0,0,0,green,white
3,American-Samoa,6,3,0,0,1,1,0,0,5,1,0,1,1,1,0,1,blue,0,0,0,0,0,0,1,1,1,0,blue,red
4,Andorra,3,1,0,0,6,0,3,0,3,1,0,1,1,0,0,0,gold,0,0,0,0,0,0,0,0,0,0,blue,red


In [6]:
# check size of dataset
flags.shape

(194, 30)

In [8]:
# check general information about dataset
flags.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194 entries, 0 to 193
Data columns (total 30 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   name            194 non-null    object
 1   landmass        194 non-null    int64 
 2   zone            194 non-null    int64 
 3   area            194 non-null    int64 
 4   population      194 non-null    int64 
 5   language        194 non-null    int64 
 6   religion        194 non-null    int64 
 7   num_bars        194 non-null    int64 
 8   num_stripes     194 non-null    int64 
 9   num_colors      194 non-null    int64 
 10  red             194 non-null    int64 
 11  green           194 non-null    int64 
 12  blue            194 non-null    int64 
 13  gold            194 non-null    int64 
 14  white           194 non-null    int64 
 15  black           194 non-null    int64 
 16  orange          194 non-null    int64 
 17  mainhue         194 non-null    object
 18  num_circle

In [10]:
# check general statistical information
flags.describe()

,landmass,zone,area,population,language,religion,num_bars,num_stripes,num_colors,red,green,blue,gold,white,black,orange,num_circles,num_crosses,num_saltires,num_quarters,num_sunstars,crescent,triangle,icon,animate,text
count,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000,194.000000
mean,3.572165,2.211340,700.046392,23.268041,5.340206,2.190722,0.453608,1.551546,3.463918,0.788660,0.469072,0.510309,0.469072,0.752577,0.268041,0.134021,0.170103,0.149485,0.092784,0.149485,1.386598,0.056701,0.139175,0.252577,0.201031,0.082474
std,1.553018,1.308274,2170.927932,91.934085,3.496517,2.061167,1.038339,2.328005,1.300154,0.409315,0.500334,0.501187,0.500334,0.432631,0.444085,0.341556,0.463075,0.385387,0.290879,0.435860,4.396186,0.231869,0.347025,0.435615,0.401808,0.275798
min,1.000000,1.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,3.000000,1.000000,9.000000,0.000000,2.000000,1.000000,0.000000,0.000000,3.000000,1.000000,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
50%,4.000000,2.000000,111.000000,4.000000,6.000000,1.000000,0.000000,0.000000,3.000000,1.000000,0.000000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
75%,5.000000,4.000000,471.250000,14.000000,9.000000,4.000000,0.000000,3.000000,4.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.750000,0.000000,0.000000
max,6.000000,4.000000,22402.000000,1008.000000,10.000000,7.000000,5.000000,14.000000,8.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,4.000000,2.000000,1.000000,4.000000,50.000000,1.000000,1.000000,1.000000,1.000000,1.000000


## GroupBy Mechanics
### Basic Grouping

Hadley Wickham, an author of many popular packages for the R programming language, coined the term split-apply-combine for describing group operations. 

- First, the data is split into groups.
- Second, a function is applied to each group
- Finally, the results are combined into a result object

Here is a mockup of a simple group aggregation.

<img src='split_apply_combine.JPG'>

*Hadley Wickham, an author of many popular packages for the R programming language, coined the term split-apply-combine for describing group operations.*

To get started, let's create a small dataset.

In [12]:
# create sample dataset 
df = pd.DataFrame({'studio_key' : ['Marvel', 'Marvel', 'DC', 'DC', 'Marvel'],
     'department_key' : ['Production', 'Advertising', 'Production', 'Advertising', 'Production'],
     'data1' : [10,6,2,7,5],
     'data2' : [-1,4,-6,5,11]})
df

,studio_key,department_key,data1,data2
0,Marvel,Production,10,-1
1,Marvel,Advertising,6,4
2,DC,Production,2,-6
3,DC,Advertising,7,5
4,Marvel,Production,5,11


Suppose we wanted to compute the mean of `data1` grouped by the `studio_key` column.

In [14]:
# create Series GroupBy object using 'studio_key'
grouped = df['data1'].groupby(df['studio_key'])
grouped

Now we can simply call the 'mean' method on the GroupBy object

In [16]:
# produces new Series
grouped.mean()

studio_key
DC        4.5
Marvel    7.0
Name: data1, dtype: float64

In [18]:
# chaining it all together
grouped = df['data1'].groupby(df['studio_key']).mean()
grouped

studio_key
DC        4.5
Marvel    7.0
Name: data1, dtype: float64

We can also easily pass multiple keys to be used by the GroupBy object. This actually creates a multi-index Series.

In [20]:
# passing multiple Series as a list
means = df['data1'].groupby([df['studio_key'], df['department_key']]).mean()
means

studio_key  department_key
DC          Advertising       7.0
            Production        2.0
Marvel      Advertising       6.0
            Production        7.5
Name: data1, dtype: float64

Remember that you can use `unstack()` to produce a DataFrame

In [24]:
# using unstack() to create a DataFrame
means.unstack()
print(np.__version__)

1.26.4


Frequently the grouping information is found in the same DataFrame as the data you want to work on. In that case, you can pass column names  as the group keys.

In [26]:
# passing DataFrame column names as group keys
df.groupby('studio_key').mean(numeric_only=True)

# df['data1'].groupby('studio_key').mean() #produces error
# df['data1'].groupby(df['studio_key']).mean() #produces Series

,data1,data2
studio_key,,
DC,4.5,-0.500000
Marvel,7.0,4.666667


Notice that there is no `department_key` in the above result. Since that column is not numeric, it is excluded from the result but you must pass the `numeric_only=True` as a parameter or you will get an error message. By default, all of the numeric columns are aggregated, though it is possible to filter down to a subset as we will soon see.

A very useful GroupBy method is `size`, which returns a Series containing group sizes.

In [28]:
# show group sizes
df.groupby(['studio_key', 'department_key']).size()

studio_key  department_key
DC          Advertising       1
            Production        1
Marvel      Advertising       1
            Production        2
dtype: int64

### Student Practice
Try to perform the following tasks on the `flags` dataset. Then check your answers as I walk through the solutions. 

**Exercise:** Instantiate (create) a SeriesGroupBy object called `grouped_flags` that selects `population` from the `flags` dataset and groups it by `landmass`

*Note: landmass: 1=N.America, 2=S.America, 3=Europe, 4=Africa, 5=Asia, 6=Oceania*

In [34]:
### ENTER CODE HERE ###
#grouped = df['data1'].groupby(df['studio_key'])

grouped_flags = flags['population'].groupby(flags['landmass'])

grouped_flags

**Exercise:** Using the `grouped_flags` object, what is the average population by landmass? What is the minimum population by landmass? What is the maximum population by landmass?

In [38]:
### ENTER CODE HERE ###
print(grouped_flags.mean())
print(grouped_flags.min())
print(grouped_flags.max())

landmass
1    12.290323
2    15.705882
3    13.857143
4     8.788462
5    69.179487
6    11.300000
Name: population, dtype: float64
landmass
1    0
2    0
3    0
4    0
5    0
6    0
Name: population, dtype: int64
landmass
1     231
2     119
3      61
4      56
5    1008
6     157
Name: population, dtype: int64


**Exercise:** Instantiate an object called `flag_means` that selects the `population` and groups it by `zone`, then by `landmass` and calculates the mean of the population.

*Note: zone: Geographic quadrant, based on Greenwich and the Equator (1=NE, 2=SE, 3=SW, 4=NW)*

In [46]:
### ENTER CODE HERE ###
flag_means = flags['population'].groupby([flags['zone'],flags['landmass']]).mean()
flag_means

zone  landmass
1     3           13.500000
      4           12.789474
      5           69.179487
      6            9.600000
2     4            7.315789
      6           17.800000
3     2           22.000000
      4            0.000000
      6            0.000000
4     1           12.290323
      2            6.714286
      3           15.285714
      4            5.769231
Name: population, dtype: float64

**Exercise:** Turn `flag_means` into a DataFrame with `zone` as the rows and `landmass` as the columns. You should be able to do this using one pandas method. Take note of the missing values in the new DataFrame.

In [48]:
### ENTER CODE HERE ###
flag_means.unstack()

landmass,1,2,3,4,5,6
zone,,,,,,
1,NaN,NaN,13.500000,12.789474,69.179487,9.6
2,NaN,NaN,NaN,7.315789,NaN,17.8
3,NaN,22.000000,NaN,0.000000,NaN,0.0
4,12.290323,6.714286,15.285714,5.769231,NaN,NaN


**Exercise:** Group the entire `flags` dataset by `landmass` and compute the median of each numeric column.

In [54]:
### ENTER CODE HERE ###
flags.groupby(flags['landmass']).mean(numeric_only=True)
#flags.groupby('landmass').median(numeric_only=True)

,zone,area,population,language,religion,num_bars,num_stripes,num_colors,red,green,blue,gold,white,black,orange,num_circles,num_crosses,num_saltires,num_quarters,num_sunstars,crescent,triangle,icon,animate,text
landmass,,,,,,,,,,,,,,,,,,,,,,,,,
1,4.000000,779.451613,12.290323,1.677419,0.870968,0.580645,1.451613,3.870968,0.677419,0.419355,0.774194,0.516129,0.838710,0.354839,0.193548,0.193548,0.225806,0.193548,0.193548,2.806452,0.000000,0.258065,0.290323,0.419355,0.161290
2,3.411765,1211.117647,15.705882,2.352941,0.411765,0.352941,2.176471,3.705882,0.764706,0.411765,0.705882,0.647059,0.823529,0.235294,0.058824,0.176471,0.058824,0.058824,0.411765,2.117647,0.000000,0.117647,0.176471,0.176471,0.176471
3,1.600000,139.514286,13.857143,5.142857,1.742857,0.628571,1.285714,3.000000,0.828571,0.200000,0.514286,0.428571,0.771429,0.171429,0.057143,0.028571,0.285714,0.028571,0.028571,0.142857,0.000000,0.028571,0.257143,0.114286,0.028571
4,2.153846,534.307692,8.788462,6.769231,3.423077,0.673077,1.942308,3.519231,0.807692,0.826923,0.307692,0.519231,0.538462,0.326923,0.134615,0.115385,0.038462,0.057692,0.057692,0.576923,0.076923,0.134615,0.173077,0.134615,0.038462
5,1.000000,1222.897436,69.179487,8.487179,3.307692,0.179487,1.666667,3.153846,0.820513,0.435897,0.333333,0.256410,0.820513,0.282051,0.179487,0.358974,0.025641,0.025641,0.102564,1.153846,0.179487,0.102564,0.333333,0.153846,0.102564
6,2.000000,534.850000,11.300000,4.050000,1.150000,0.000000,0.400000,3.900000,0.800000,0.200000,0.800000,0.600000,0.950000,0.150000,0.150000,0.150000,0.400000,0.300000,0.400000,3.300000,0.000000,0.250000,0.300000,0.300000,0.050000


**Exercise:** How many countries are represented in each group if you group by `landmass` and then `zone`.

In [60]:
### ENTER CODE HERE ###
flags.groupby(['landmass', 'zone']).size()

landmass  zone
1         4       31
2         3       10
          4        7
3         1       28
          4        7
4         1       19
          2       19
          3        1
          4       13
5         1       39
6         1        5
          2       10
          3        5
dtype: int64

### Iterating Over Groups

The GroupBy object supports iteration, generating a sequence of 2-tuples containing the group name along with the chunk of data.

Let's remind ourselves of our sample dataset.

In [62]:
# view DataFrame
df

,studio_key,department_key,data1,data2
0,Marvel,Production,10,-1
1,Marvel,Advertising,6,4
2,DC,Production,2,-6
3,DC,Advertising,7,5
4,Marvel,Production,5,11


In [64]:
# reminder: this creates a DataFrameGroupBy object
df.groupby('studio_key')

You can iterate through this object to find the name and group data.

In [66]:
# loop through object
for name, group in df.groupby('studio_key'):
    print(name)
    print(group)
    print('----')

DC
  studio_key department_key  data1  data2
2         DC     Production      2     -6
3         DC    Advertising      7      5
----
Marvel
  studio_key department_key  data1  data2
0     Marvel     Production     10     -1
1     Marvel    Advertising      6      4
4     Marvel     Production      5     11
----


In the case of multiple keys, the first element in the tuple will be a tuple of key values.

In [68]:
# loop through object
for (key1, key2), group in df.groupby(['studio_key', 'department_key']):
    print(f'key1: {key1}')
    print(f'key2: {key2}')
    print(group)
    print('----') 

key1: DC
key2: Advertising
  studio_key department_key  data1  data2
3         DC    Advertising      7      5
----
key1: DC
key2: Production
  studio_key department_key  data1  data2
2         DC     Production      2     -6
----
key1: Marvel
key2: Advertising
  studio_key department_key  data1  data2
1     Marvel    Advertising      6      4
----
key1: Marvel
key2: Production
  studio_key department_key  data1  data2
0     Marvel     Production     10     -1
4     Marvel     Production      5     11
----


By default `groupby` groups on axis=0, but you can group on any of the other axes. For example, we could group the columns of our example `df` here by `dtype` like so:

<span style="color:red">Note: The below code has now been deprecated. For axis=1, do frame.T.groupby(...) instead. See third code block below for updated code.</span>

In [70]:
# check data types
df.dtypes

studio_key        object
department_key    object
data1              int64
data2              int64
dtype: object

In [ ]:
### DEPRECATED ###

# print data type and respective group data
#grouped = df.groupby(df.dtypes, axis=1)
#for dtype, group in grouped:
#    print(dtype)
#    print(group)
#    print('----')

In [72]:
### UPDATED ###

# print data type and respective group data
grouped = df.T.groupby(df.dtypes)
for dtype, group in grouped:
    print(dtype)
    print(group.T)
    print('----')

int64
  data1 data2
0    10    -1
1     6     4
2     2    -6
3     7     5
4     5    11
----
object
  studio_key department_key
0     Marvel     Production
1     Marvel    Advertising
2         DC     Production
3         DC    Advertising
4     Marvel     Production
----


### Selecting a Column or Subset of Columns

In [74]:
# groupby studio_key, average of 'data1', returns Series
df['data1'].groupby(df['studio_key']).mean()

studio_key
DC        4.5
Marvel    7.0
Name: data1, dtype: float64

In [76]:
# syntactic sugar for above
df.groupby('studio_key')['data1'].mean()

studio_key
DC        4.5
Marvel    7.0
Name: data1, dtype: float64

In [80]:
# groupby 'studio_key', average of data2, returns DataFrame
df[['data2']].groupby(df['studio_key']).mean()

,data2
studio_key,
DC,-0.500000
Marvel,4.666667


In [78]:
# syntactic sugar for above
df.groupby('studio_key')[['data2']].mean()

,data2
studio_key,
DC,-0.500000
Marvel,4.666667


Especially for large datasets, it may be desirable to aggregate only a few columns. For example, in the preceding dataset, to compute means for just the data2 column and get the result as a DataFrame, we could write:

In [82]:
# grouped DataFram
df.groupby(['studio_key', 'department_key'])[['data2']].mean()

data2
studio_key department_key       
DC         Advertising       5.0
           Production       -6.0
Marvel     Advertising       4.0
           Production        5.0

In [84]:
# grouped Series
df.groupby(['studio_key', 'department_key'])['data2'].mean()

studio_key  department_key
DC          Advertising       5.0
            Production       -6.0
Marvel      Advertising       4.0
            Production        5.0
Name: data2, dtype: float64

### Grouping with Dictionaries and Series

To see different ways to work with grouping, let's create a DataFrame of student grades that have taken two courses.

In [86]:
# create another sample DataFrame
rng = np.random.default_rng(12345)

students = pd.DataFrame(rng.integers(low=80, high=100, size=(5, 5)),
                        columns=[1, 2, 3, 4, 5],
                        index=['Joe', 'Steve', 'Beth', 'Jim', 'Sue'])
# add some NA values
students.iloc[2:3,[1,2]] = np.nan 

students

,1,2,3,4,5
Joe,93,84.0,95.0,86,84
Steve,95,92.0,93.0,99,87
Beth,96,NaN,NaN,91,84
Jim,83,84.0,93.0,92,98
Sue,94,84.0,98.0,98,94


Let's say that we want to map course names to the specific quiz. We can do this with a mapping and then groupby this dictionary mapping.

In [88]:
# create mapping
mapping = {1:670,2:670,3:520,4:520,5:670,6:680}

In [ ]:
### DEPRECATED ###

# passing the mapping to the groupby object
#by_column = students.groupby(mapping, axis=1)

# summing by the grouped mapping, notice the unused mapping is OK
#by_column.mean()

In [90]:
### UPDATED CODE ###

# passing the mapping to the groupby object
by_column = students.T.groupby(mapping)

# summing by the grouped mapping, notice the unused mapping is OK
by_column.mean().T

,520,670
Joe,90.5,87.000000
Steve,96.0,91.333333
Beth,91.0,90.000000
Jim,92.5,88.333333
Sue,98.0,90.666667


We can also work with Series:

In [92]:
# create Series of
map_series = pd.Series(mapping)
map_series

1    670
2    670
3    520
4    520
5    670
6    680
dtype: int64

In [ ]:
### DEPRECATED ###

# pass Series to groupby object
#students.groupby(map_series, axis=1).count()

In [94]:
### UPDATED ###

# pass Series to groupby object
students.T.groupby(map_series).count().T

,520,670
Joe,2,3
Steve,2,3
Beth,1,2
Jim,2,3
Sue,2,3


### Grouping with Functions

Any function passed as a group key will be called once per index value.

As an example, let's say we wanted to group students based on how many letters were in their name and find their median score. (Why we would ever want to do this? Who knows. Just go along with me here.)

In [96]:
# group by index length (in this case student name)
students.groupby(len).median()

,1,2,3,4,5
3,93.0,84.0,95.0,92.0,94.0
4,96.0,NaN,NaN,91.0,84.0
5,95.0,92.0,93.0,99.0,87.0


In [98]:
# let's rename the columns
students = students.rename(columns=mapping)
students

,670,670,520,520,670
Joe,93,84.0,95.0,86,84
Steve,95,92.0,93.0,99,87
Beth,96,NaN,NaN,91,84
Jim,83,84.0,93.0,92,98
Sue,94,84.0,98.0,98,94


In [100]:
# create a another key list
key_list = ['MA','NY','NY','MA','NY']

# groupby function, then by key_list
students.groupby([len, key_list]).mean()

670   670   520   520   670
3 MA  88.0  84.0  94.0  89.0  91.0
  NY  94.0  84.0  98.0  98.0  94.0
4 NY  96.0   NaN   NaN  91.0  84.0
5 NY  95.0  92.0  93.0  99.0  87.0

### Grouping by Index Levels

You can easily aggregate using one of the levels of a multi-index. Let's add a `gender` column to our `students` data and create a multi-index DataFrame.

In [102]:
# adding gender column
students['gender'] = ['M','M','F','M','F']
students

,670,670,520,520,670,gender
Joe,93,84.0,95.0,86,84,M
Steve,95,92.0,93.0,99,87,M
Beth,96,NaN,NaN,91,84,F
Jim,83,84.0,93.0,92,98,M
Sue,94,84.0,98.0,98,94,F


In [104]:
# creating multi-index
students = students.reset_index().set_index(['index','gender'])
students

,,670,670,520,520,670
index,gender,,,,,
Joe,M,93,84.0,95.0,86,84
Steve,M,95,92.0,93.0,99,87
Beth,F,96,NaN,NaN,91,84
Jim,M,83,84.0,93.0,92,98
Sue,F,94,84.0,98.0,98,94


In [106]:
# grouping by 'gender' index and counting number of quizzes taken by gender
students.groupby(level='gender').count()

,670,670,520,520,670
gender,,,,,
F,2,1,1,2,2
M,3,3,3,3,3


### Student Practice

Try to perform the following tasks on the `flags` dataset. Then check your answers as I walk through the solutions. 

In [108]:
# let's remind oursleves of our DataFrame
flags.head()

,name,landmass,zone,area,population,language,religion,num_bars,num_stripes,num_colors,red,green,blue,gold,white,black,orange,mainhue,num_circles,num_crosses,num_saltires,num_quarters,num_sunstars,crescent,triangle,icon,animate,text,topleft_color,botright_color
0,Afghanistan,5,1,648,16,10,2,0,3,5,1,1,0,1,1,1,0,green,0,0,0,0,1,0,0,1,0,0,black,green
1,Albania,3,1,29,3,6,6,0,0,3,1,0,0,1,0,1,0,red,0,0,0,0,1,0,0,0,1,0,red,red
2,Algeria,4,1,2388,20,8,2,2,0,3,1,1,0,0,1,0,0,green,0,0,0,0,1,1,0,0,0,0,green,white
3,American-Samoa,6,3,0,0,1,1,0,0,5,1,0,1,1,1,0,1,blue,0,0,0,0,0,0,1,1,1,0,blue,red
4,Andorra,3,1,0,0,6,0,3,0,3,1,0,1,1,0,0,0,gold,0,0,0,0,0,0,0,0,0,0,blue,red


**Exercise:** What is the sum of the `num_crosses` grouped by `religion`?

*Note: religion: 0=Catholic, 1=Other Christian, 2=Muslim, 3=Buddhist, 4=Hindu, 5=Ethnic, 6=Marxist, 7=Others*

In [110]:
### ENTER CODE HERE ###
#flags['population'].groupby([flags['zone'],flags['landmass']]).mean()
flags['num_crosses'].groupby(flags['religion']).sum()

religion
0     2
1    26
2     0
3     1
4     0
5     0
6     0
7     0
Name: num_crosses, dtype: int64

**Exercise:** What is the sum of `crescent` grouped by `religion`?

In [112]:
### ENTER CODE HERE ###
flags['crescent'].groupby(flags['religion']).sum()

religion
0    0
1    0
2    8
3    1
4    1
5    0
6    1
7    0
Name: crescent, dtype: int64

**Exercise:** What are the total value counts of all the colors in `mainhue` grouped by `religion`?

In [114]:
### ENTER CODE HERE ###
flags['mainhue'].groupby(flags['religion']).value_counts()

religion  mainhue
0         red        15
          blue        9
          gold        6
          white       6
          green       3
          black       1
1         blue       24
          red        16
          white       9
          green       6
          gold        3
          black       1
          orange      1
2         red        15
          green      12
          gold        3
          black       2
          blue        2
          brown       1
          orange      1
3         red         4
          blue        1
          gold        1
          orange      1
          white       1
4         brown       1
          green       1
          orange      1
          red         1
5         red        10
          green       8
          gold        5
          blue        2
          black       1
          white       1
6         red        10
          blue        2
          white       2
          gold        1
7         white       3
          green       

**Exercise:** What is the maximum `area` for each group that is grouped by `zone` and then `religion`? What is the minimum area?

In [118]:
### ENTER CODE HERE ###
#flag_means = flags['population'].groupby([flags['zone'],flags['landmass']]).mean()
flags['area'].groupby([flags['zone'],flags['religion']]).max()

zone  religion
1     0             547
      1            1222
      2            2506
      3             678
      4            3268
      5            1284
      6           22402
      7             372
2     1            7690
      2            1904
      4               2
      5            1247
3     0            8512
      1              12
4     0            1973
      1            9976
      2            1240
      4             215
      5             323
      6             115
Name: area, dtype: int64

**Exercise:** Let's try to determine if there are more colors or shapes on the country flags.
1. Create a subset of the `flags` data and call it `flags_subset`. This subset should include the following attributes: 'red', 'green', 'blue', 'gold', 'white', 'black', 'orange', 'num_circles', 'num_crosses', 'num_saltires', 'num_sunstars', 'crescent',  and 'triangle'
2. Create a dictionary that maps all colors to the string `color` and all shapes to the string `shape`
3. Use the mapping dictionary from step 2 to calculate the sum of the colors and shapes for each instance.
4. *Bonus:* Sum up all the colors and shapes for all instances to determine if there are more colors or shapes on all the flags.

<Span style="color:red">Note: The solution shown in the video is deprecated. Refer to the updated code in the previous sections related to using `axis=1` to help you update the original solution.</span>

In [128]:
### ENTER CODE HERE ###
flags_subset=flags[['red', 'green', 'blue', 'gold', 'white', 'black', 'orange', 'num_circles', 'num_crosses', 
                    'num_saltires', 'num_sunstars', 'crescent', 'triangle']]
##create a mapping dictionary
mapping = {'red': 'color', 'green': 'color', 'blue': 'color', 'gold': 'color', 'white': 'color', 'black': 'color', 'orange': 'color',
           'num_circles': 'shape', 'num_crosses': 'shape', 'num_saltires': 'shape', 'crescent': 'shape', 'triangle': 'shape'}

#by_column = flag_subs

In [130]:
by_column = flags_subset.groupby(mapping, axis=1)
by_column.sum()

/var/folders/bn/mf995m8j15s64ffx8y6tfrrjfpjs54/T/ipykernel_50068/4220433644.py:1: FutureWarning: DataFrame.groupby with axis=1 is deprecated. Do `frame.T.groupby(...)` without axis instead.
  by_column = flags_subset.groupby(mapping, axis=1)


,color,shape
0,5,0
1,3,0
2,3,1
3,5,1
4,3,0
...,...,...
189,3,0
190,4,0
191,4,1
192,4,0


In [132]:
by_column.sum().sum()

color    658
shape    118
dtype: int64

### Data Aggregation

You can use aggregations of your own devising and additionally call any method that is also defined on the grouped object. 

Let's look at another simple example. First, let's create a similar DataFrame to our `students` that represents the grades of five quizzes these students received in one class.

In [136]:
# create another sample DataFrame
rng = np.random.default_rng(54321)

quiz_df = pd.DataFrame(rng.integers(low=70, high=100, size=(8, 5)),
                        columns=[1, 2, 3, 4, 5],
                        index=['Joe','Steve','Beth','Jim','Sue','James','Amy','Monika'])

# add a gender column for grouping
quiz_df['gender'] = ['M','M','F','M','F','M','F','F']
print(quiz_df)

         1   2   3   4   5 gender
Joe     70  91  97  95  89      M
Steve   84  90  70  82  83      M
Beth    94  73  97  91  91      F
Jim     81  76  84  78  87      M
Sue     76  91  84  89  77      F
James   70  71  92  95  92      M
Amy     89  70  98  94  96      F
Monika  90  91  84  83  80      F


In [138]:
# groupby gender
grouped = quiz_df.groupby('gender')

In [142]:
# use agg and pass 'mean'
# notice you pass this as a string
grouped.agg('mean')

# notice that this is the same as passing the following:
#grouped.mean()

,1,2,3,4,5
gender,,,,,
F,87.25,81.25,90.75,89.25,86.00
M,76.25,82.00,85.75,87.50,87.75


Now comes the fun part. Let's say that you wanted to know the range of the top score and bottom score for each quiz broken down by gender. We can create our own custom function to do this.

In [144]:
# create a simple custom function
def range_scores(arr):
    return arr.max() - arr.min()

# pass function to agg
grouped.agg(range_scores)

,1,2,3,4,5
gender,,,,,
F,18,21,14,11,19
M,14,20,27,17,9


Note that you can also pass the describe method to a grouped object.

In [146]:
grouped[1].describe()

,count,mean,std,min,25%,50%,75%,max
gender,,,,,,,,
F,4.0,87.25,7.804913,76.0,85.75,89.5,91.00,94.0
M,4.0,76.25,7.320064,70.0,70.00,75.5,81.75,84.0


Let's add another column that lists the student's level.

In [148]:
quiz_df['level'] = ['Senior','Junior','Senior','Senior','Junior','Junior','Senior','Junior']
quiz_df

,1,2,3,4,5,gender,level
Joe,70,91,97,95,89,M,Senior
Steve,84,90,70,82,83,M,Junior
Beth,94,73,97,91,91,F,Senior
Jim,81,76,84,78,87,M,Senior
Sue,76,91,84,89,77,F,Junior
James,70,71,92,95,92,M,Junior
Amy,89,70,98,94,96,F,Senior
Monika,90,91,84,83,80,F,Junior


Next, let's add an average quiz score for each student.

In [150]:
quiz_df['avg'] = quiz_df[[1,2,3,4,5]].mean(axis=1)
quiz_df

,1,2,3,4,5,gender,level,avg
Joe,70,91,97,95,89,M,Senior,88.4
Steve,84,90,70,82,83,M,Junior,81.8
Beth,94,73,97,91,91,F,Senior,89.2
Jim,81,76,84,78,87,M,Senior,81.2
Sue,76,91,84,89,77,F,Junior,83.4
James,70,71,92,95,92,M,Junior,84.0
Amy,89,70,98,94,96,F,Senior,89.4
Monika,90,91,84,83,80,F,Junior,85.6


Now, let's group by student level and then by gender. We will select only the `avg` column and see what the mean score is for the respective groupings.

In [156]:
# groupby level and gender
grouped = quiz_df.groupby(['level','gender'])

# selecting only the avg column
grouped_avg = grouped['avg']

# aggregating the mean
grouped_avg.agg('mean')


level   gender
Junior  F         84.5
        M         82.9
Senior  F         89.3
        M         84.8
Name: avg, dtype: float64

In [152]:
# same as above
grouped['avg'].mean()

gender
F    86.90
M    83.85
Name: avg, dtype: float64

If you pass a list of functions or function names instead, you get back a DataFrame with column names taken from the functions:

In [158]:
grouped_avg.agg(['mean','std',range_scores])

mean       std  range_scores
level  gender                              
Junior F       84.5  1.555635           2.2
       M       82.9  1.555635           2.2
Senior F       89.3  0.141421           0.2
       M       84.8  5.091169           7.2

You can also create a list of functions and pass this list to `agg`

In [160]:
# list of functions
functions = ['mean','std',range_scores]

grouped_avg.agg(functions)

mean       std  range_scores
level  gender                              
Junior F       84.5  1.555635           2.2
       M       82.9  1.555635           2.2
Senior F       89.3  0.141421           0.2
       M       84.8  5.091169           7.2

You can also change the name of the column when you aggregate like this:

In [162]:
grouped_avg.agg([('Average', 'mean'), ('Std Dev', 'std'), ('Range', range_scores)])

Average   Std Dev  Range
level  gender                          
Junior F          84.5  1.555635    2.2
       M          82.9  1.555635    2.2
Senior F          89.3  0.141421    0.2
       M          84.8  5.091169    7.2

Finally, let's say that we want to apply a different function to separate columns of the DataFrame. You can pass a dictionary like this:

In [164]:
grouped.agg({1:'mean', 2:'median', 3:'count'})

1     2  3
level  gender               
Junior F       83.0  91.0  2
       M       77.0  80.5  2
Senior F       91.5  71.5  2
       M       75.5  83.5  2

### Apply (split-apply-combine)

The most general-purpose GroupBy method is `apply`. `apply` splits the object being manipulated into pieces, invokes the passed function on each piece, and then attempts to concatenate the pieces together.

Let's start by creating a new sample DataFrame of flights from Vienna to Charlotte.

In [166]:
# Create a random number generator
rng = np.random.default_rng(12345)

# Create sample flights data
flights = pd.DataFrame({
    'airline': ['Delta', 'Delta', 'Delta', 'Delta', 'Delta', 
                'United', 'United', 'United', 'United', 
                'Lufthansa', 'Lufthansa', 'Lufthansa', 'Lufthansa', 'Lufthansa', 'Lufthansa',
                'British Airways', 'British Airways', 'British Airways'],
    'price': rng.integers(600, 1000, size=18),
    'time': rng.integers(8, 16, size=18)
})

flights

,airline,price,time
0,Delta,879,12
1,Delta,690,15
2,Delta,915,13
3,Delta,726,9
4,Delta,681,15
5,United,918,15
6,United,857,13
7,United,870,13
8,United,995,9
9,Lufthansa,756,8


Now, let's create a custom function that selects `n` rows with the lowest values in a particular column.

In [168]:
# create custom function
def best(df, n=3, column='price'):
    return df.sort_values(by=column)[:n]

In [170]:
# test on full data
best(flights)

,airline,price,time
15,British Airways,674,13
4,Delta,681,15
14,Lufthansa,685,11


Now, let's group by `airlines` and call `apply` with this function.

In [ ]:
### DEPRECATED ###

#flights.groupby('airline').apply(best)

In [174]:
### UPDATED ###

flights.groupby('airline')[['airline','price','time']].apply(best)

airline  price  time
airline                                         
British Airways 15  British Airways    674    13
                16  British Airways    691     9
                17  British Airways    869    10
Delta           4             Delta    681    15
                1             Delta    690    15
                3             Delta    726     9
Lufthansa       14        Lufthansa    685    11
                11        Lufthansa    733    11
                9         Lufthansa    756     8
United          6            United    857    13
                7            United    870    13
                5            United    918    15

You can also add other arguments in the `apply` method. What if we wanted only the two shortest flights by airline?

In [ ]:
### DEPRECATED ###

# two shortest flights grouped by airline
#flights.groupby('airline').apply(best, n=2, column='time')

In [176]:
### UPDATED ###

# two shortest flights grouped by airline
flights.groupby('airline')[['airline','price','time']].apply(best, n=2, column='time')

airline  price  time
airline                                         
British Airways 16  British Airways    691     9
                17  British Airways    869    10
Delta           3             Delta    726     9
                0             Delta    879    12
Lufthansa       9         Lufthansa    756     8
                12        Lufthansa    827     8
United          8            United    995     9
                6            United    857    13

### Student Practice

Try to perform the following tasks on the `flags` dataset. Then check your answers as I walk through the solutions. 

In [178]:
flags.head()

,name,landmass,zone,area,population,language,religion,num_bars,num_stripes,num_colors,red,green,blue,gold,white,black,orange,mainhue,num_circles,num_crosses,num_saltires,num_quarters,num_sunstars,crescent,triangle,icon,animate,text,topleft_color,botright_color
0,Afghanistan,5,1,648,16,10,2,0,3,5,1,1,0,1,1,1,0,green,0,0,0,0,1,0,0,1,0,0,black,green
1,Albania,3,1,29,3,6,6,0,0,3,1,0,0,1,0,1,0,red,0,0,0,0,1,0,0,0,1,0,red,red
2,Algeria,4,1,2388,20,8,2,2,0,3,1,1,0,0,1,0,0,green,0,0,0,0,1,1,0,0,0,0,green,white
3,American-Samoa,6,3,0,0,1,1,0,0,5,1,0,1,1,1,0,1,blue,0,0,0,0,0,0,1,1,1,0,blue,red
4,Andorra,3,1,0,0,6,0,3,0,3,1,0,1,1,0,0,0,gold,0,0,0,0,0,0,0,0,0,0,blue,red


**Exercise:** Group the data by `zone` and determine the difference between the zone's largest area and its smallest area. Do this by creating a custom function and passing it to the `agg` method.

In [194]:
### ENTER CODE HERE ###
#def diff(flags, column='area'):
  #  return flags['area'].max() - flags['area'].min()

#diff(flags)

##prof solution
grouped = flags.groupby('zone')
grouped_zone_area = grouped['area']

#create function to return range

def col_range(arr):
    return arr.max()-arr.min()

##pass custom function to agg

grouped_zone_area.agg(col_range)

zone
1    22402
2     7690
3     8512
4     9976
Name: area, dtype: int64

**Exercise:** Add a new column called `pop_den` to the DataFrame that represents the respective country's population density. Population density is defined as the population divided by the area.

In [196]:
### ENTER CODE HERE ###

flags['pop_den']=(flags['population']/flags['area']).round(2)
flags

,name,landmass,zone,area,population,language,religion,num_bars,num_stripes,num_colors,red,green,blue,gold,white,black,orange,mainhue,num_circles,num_crosses,num_saltires,num_quarters,num_sunstars,crescent,triangle,icon,animate,text,topleft_color,botright_color,pop_density,pop_den
0,Afghanistan,5,1,648,16,10,2,0,3,5,1,1,0,1,1,1,0,green,0,0,0,0,1,0,0,1,0,0,black,green,0.02,0.02
1,Albania,3,1,29,3,6,6,0,0,3,1,0,0,1,0,1,0,red,0,0,0,0,1,0,0,0,1,0,red,red,0.10,0.10
2,Algeria,4,1,2388,20,8,2,2,0,3,1,1,0,0,1,0,0,green,0,0,0,0,1,1,0,0,0,0,green,white,0.01,0.01
3,American-Samoa,6,3,0,0,1,1,0,0,5,1,0,1,1,1,0,1,blue,0,0,0,0,0,0,1,1,1,0,blue,red,NaN,NaN
4,Andorra,3,1,0,0,6,0,3,0,3,1,0,1,1,0,0,0,gold,0,0,0,0,0,0,0,0,0,0,blue,red,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
189,Western-Samoa,6,3,3,0,1,1,0,0,3,1,0,1,0,1,0,0,red,0,0,0,1,5,0,0,0,0,0,blue,red,0.00,0.00
190,Yugoslavia,3,1,256,22,6,6,0,3,4,1,0,1,1,1,0,0,red,0,0,0,0,1,0,0,0,0,0,blue,red,0.09,0.09
191,Zaire,4,2,905,28,10,5,0,0,4,1,1,0,1,0,0,1,green,1,0,0,0,0,0,0,1,1,0,green,green,0.03,0.03
192,Zambia,4,2,753,6,10,5,3,0,4,1,1,0,0,0,1,1,green,0,0,0,0,0,0,0,0,1,0,green,brown,0.01,0.01


**Exercise:** Group by `landmass` and determine the mean, median, standard deviation and range for the population density column. Call the columns 'Avg', 'Median', 'Std Dev' and 'Range' respectively.

In [198]:
### ENTER CODE HERE ###
##flights.groupby('airline')[['airline','price','time']].apply(best, n=2, column='time')
##group landmass
grouped = flags.groupby(['landmass'])

#select pop den
grouped_pd = grouped['pop_den']

#aggr the respetive function

grouped_pd.agg([('Avg', 'mean'), ('Median', 'median'), ('Std Dev', 'std'), ('Range', col_range)])

,Avg,Median,Std Dev,Range
landmass,,,,
1,0.087500,0.04,0.101160,0.33
2,0.022941,0.01,0.046605,0.20
3,0.111786,0.10,0.088655,0.34
4,0.038980,0.02,0.076137,0.50
5,0.317368,0.05,0.921584,5.00
6,0.026667,0.00,0.049787,0.16


**Exercise:** Grouping the data by landmass, what is the max number of bars for each group, the average number of stripes, and the median value for number of colors?

In [200]:
### ENTER CODE HERE ###

grouped.agg({'num_bars': 'max', 'num_stripes': 'mean', 'num_colors': 'median'})

,num_bars,num_stripes,num_colors
landmass,,,
1,5,1.451613,3.0
2,3,2.176471,3.0
3,3,1.285714,3.0
4,3,1.942308,3.0
5,3,1.666667,3.0
6,0,0.400000,4.0


**Exercise:** 
1. Create a custom function called `top` that returns the top 2 rows with the **largest** values in the `pop_den` column. 
2. Make sure that you do not include any rows with NaNs in the `pop_den` column.
3. Setup your function arguments so that you can change the number of rows to show and which column to sort by.
4. Group the `flags` data by `landmass` and use the apply function with your custom function.

<Span style="color:red">Note: The solutions for the next three exercises shown in the video is deprecated. Include `include_groups=False` as an argument in your apply method to update the original solution.</span>

In [204]:
### ENTER CODE HERE ###

def top(df, n=2, column='pop_den'):
    return df.dropna(subset=['pop_den']).sort_values(by=column, ascending=False)[:n]

##group landmass and apply function
flags.groupby('landmass').apply(top)

/var/folders/bn/mf995m8j15s64ffx8y6tfrrjfpjs54/T/ipykernel_50068/1150416953.py:7: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  flags.groupby('landmass').apply(top)


name  landmass  zone  area  population  language  \
landmass                                                                    
1        140      Puerto-Rico         1     4     9           3         2   
         52       El-Salvador         1     4    21           5         2   
2        173  Trinidad-Tobago         2     4     5           1         1   
         133           Panama         2     4    76           2         2   
3        121      Netherlands         3     1    41          14         6   
         16           Belgium         3     1    31          10         6   
4        111        Mauritius         4     2     2           1         1   
         143           Rwanda         4     2    26           5        10   
5        78         Hong-Kong         5     1     1           5         7   
         150        Singapore         5     1     1           3         7   
6        137      Philippines         6     1   300          48        10   
         82         Indonesia         6     2  1904         157        10   

              religion  num_bars  num_stripes  num_colors  red  green  blue  \
landmass                                                                      
1        140         0         0            5           3    1      0     1   
         52          0         0            3           2    0      0     1   
2        173         1         0            0           3    1      0     0   
         133         0         0            0           3    1      0     1   
3        121         1         0            3           3    1      0     1   
         16          0         3            0           3    1      0     0   
4        111         4         0            4           4    1      1     1   
         143         5         3            0           4    1      1     0   
5        78          3         0            0           6    1      1     1   
         150         3         0            2           2    1      0     0   
6        137         0         0            0           4    1      0     1   
         82          2         0            2           2    1      0     0   

              gold  white  black  orange mainhue  num_circles  num_crosses  \
landmass                                                                     
1        140     0      1      0       0     red            0            0   
         52      0      1      0       0    blue            0            0   
2        173     0      1      1       0     red            0            0   
         133     0      1      0       0     red            0            0   
3        121     0      1      0       0     red            0            0   
         16      1      0      1       0    gold            0            0   
4        111     1      0      0       0     red            0            0   
         143     1      0      1       0     red            0            0   
5        78      1      1      0       1    blue            1            1   
         150     0      1      0       0   white            0            0   
6        137     1      1      0       0    blue            0            0   
         82      0      1      0       0     red            0            0   

              num_saltires  num_quarters  num_sunstars  crescent  triangle  \
landmass                                                                     
1        140             0             0             1         0         1   
         52              0             0             0         0         0   
2        173             0             0             0         0         1   
         133             0             4             2         0         0   
3        121             0             0             0         0         0   
         16              0             0             0         0         0   
4        111             0             0             0         0         0   
         143             0             0    

**Exercise:** Using the above custom function (`top`), return the top 3 rows with the highest `population` grouped by `zone`.

In [208]:
### ENTER CODE HERE ###
flags.groupby('zone').apply(top, n=3, column='population', include_groups=False)

name  landmass   area  population  language  religion  \
zone                                                                      
1    37          China         5   9561        1008         7         6   
     81          India         5   3268         684         6         4   
     184          USSR         5  22402         274         5         6   
2    82      Indonesia         6   1904         157        10         2   
     153  South-Africa         4   1221          29         6         1   
     191         Zaire         4    905          28        10         5   
3    23         Brazil         2   8512         119         6         0   
     8       Argentina         2   2777          28         2         0   
     9       Argentine         2   2777          28         2         0   
4    183           USA         1   9363         231         1         1   
     112        Mexico         1   1973          77         2         0   
     180            UK         3    245          56         1         1   

          num_bars  num_stripes  num_colors  red  green  blue  gold  white  \
zone                                                                         
1    37          0            0           2    1      0     0     1      0   
     81          0            3           4    0      1     1     0      1   
     184         0            0           2    1      0     0     1      0   
2    82          0            2           2    1      0     0     0      1   
     153         0            3           5    1      1     1     0      1   
     191         0            0           4    1      1     0     1      0   
3    23          0            0           4    0      1     1     1      1   
     8           0            3           2    0      0     1     0      1   
     9           0            3           3    0      0     1     1      1   
4    183         0           13           3    1      0     1     0      1   
     112         3            0           4    1      1     0     0      1   
     180         0            0           3    1      0     1     0      1   

          black  orange mainhue  num_circles  num_crosses  num_saltires  \
zone                                                                      
1    37       0       0     red            0            0             0   
     81       0       1  orange            1            0             0   
     184      0       0     red            0            0             0   
2    82       0       0     red            0            0             0   
     153      0       1  orange            0            1             1   
     191      0       1   green            1            0             0   
3    23       0       0   green            1            0             0   
     8        0       0    blue            0            0             0   
     9        0       0    blue            0            0             0   
4    183      0       0   white            0            0             0   
     112      0       1   green            0            0             0   
     180      0       0     red            0            1             1   

          num_quarters  num_sunstars  crescent  triangle  icon  animate  text  \
zone                                                                            
1    37              0             5         0         0     0        0     0   
     81              0             0         0         0     1        0     0   
     184             0             1         0         0     1        0     0   
2    82              0             0         0         0     0        0     0   
     153             0             0         0         0     0        0     0   
     191             0             0         0         0     1        1     0   
3    23              0            22         0         0     0        0     1   
     8               0             0         0         0     0        0     0   
     9       

**Exercise:** 
1. Create a second custom function called `bottom` that returns the 2 rows with the smallest values in the `pop_den` column. Do not include any rows with a `pop_den` of `0`.
2. Group the `flags` data by `landmass` and use the apply function with your custom function.

In [212]:
### ENTER CODE HERE ###
def bottom (df, n=2, column = 'pop_den'):
    return df[df[column] !=0].sort_values(by=column)[:n]

In [216]:
#group landmass and apply function

flags.groupby('landmass').apply(bottom, include_groups=False)

name  zone  area  population  language  religion  \
landmass                                                                     
1        124         Nicaragua     4   128           3         2         0   
         183               USA     4  9363         231         1         1   
2        8           Argentina     3  2777          28         2         0   
         9           Argentine     3  2777          28         2         0   
3        130            Norway     1   324           4         6         1   
         58            Finland     1   337           5         9         1   
4        2             Algeria     1  2388          20         8         2   
         5              Angola     2  1247           7        10         5   
5        155       South-Yemen     1   288           2         8         2   
         178               UAE     1    84           1         8         2   
6        123       New-Zealand     2   268           2         1         1   
         134  Papua-New-Guinea     2   463           3         1         5   

              num_bars  num_stripes  num_colors  red  green  blue  gold  \
landmass                                                                  
1        124         0            3           2    0      0     1     0   
         183         0           13           3    1      0     1     0   
2        8           0            3           2    0      0     1     0   
         9           0            3           3    0      0     1     1   
3        130         0            0           3    1      0     1     0   
         58          0            0           2    0      0     1     0   
4        2           2            0           3    1      1     0     0   
         5           0            2           3    1      0     0     1   
5        155         0            3           4    1      0     1     0   
         178         1            3           4    1      1     0     0   
6        123         0            0           3    1      0     1     0   
         134         0            0           4    1      0     0     1   

              white  black  orange mainhue  num_circles  num_crosses  \
landmass                                                               
1        124      1      0       0    blue            0            0   
         183      1      0       0   white            0            0   
2        8        1      0       0    blue            0            0   
         9        1      0       0    blue            0            0   
3        130      1      0       0     red            0            1   
         58       1      0       0   white            0            1   
4        2        1      0       0   green            0            0   
         5        0      1       0     red            0            0   
5        155      1      1       0     red            0            0   
         178      1      1       0   green            0            0   
6        123      1      0       0    blue            0            1   
         134      1      1       0   black            0            0   

              num_saltires  num_quarters  num_sunstars  crescent  triangle  \
landmass                                                                     
1        124             0             0             0         0         0   
         183             0             1            50         0         0   
2        8               0             0             0         0         0   
         9               0             0             1         0         0   
3        130             0             0             0         0         0   
         58              0             0             0         0         0   
4        2               0             0             1         1         0   
         5               0             0             1         0         0   
5        155             0             0             1         0         1   
         178   

### More Pivot Tables and Cross-Tabulation

A pivot table aggregates a table of data by one or more keys, arranging the data in a rectangle with some of the group keys along the rows and some along the columns. Pivot tables in Python with pandas are made possible through groupby combined with reshape operations utilizing hierarchical indexing.

To show this, let's add a new column to our `flights` data that shows if the flight is in the morning or afternoon/evening.

In [218]:
# Create a random number generator
rng = np.random.default_rng(12345)

# Setup list
part_of_day = ['AM', 'PM']

# Use random choice to select AM/PM for each row
time_of_day = rng.choice(part_of_day, size=18)
time_of_day

array(['PM', 'AM', 'PM', 'AM', 'AM', 'PM', 'PM', 'PM', 'PM', 'AM', 'PM',
       'AM', 'PM', 'PM', 'AM', 'AM', 'AM', 'PM'], dtype='<U2')

In [220]:
# add new column to data
flights['time_of_day'] = time_of_day
flights

,airline,price,time,time_of_day
0,Delta,879,12,PM
1,Delta,690,15,AM
2,Delta,915,13,PM
3,Delta,726,9,AM
4,Delta,681,15,AM
5,United,918,15,PM
6,United,857,13,PM
7,United,870,13,PM
8,United,995,9,PM
9,Lufthansa,756,8,AM


Suppose you wanted to compute a table of price and time averages arranged by airline and time of day.

In [222]:
# average of price/time by airline/time of day
flights.pivot_table(index=['airline','time_of_day'])

##airline is index

price  time
airline         time_of_day                  
British Airways AM           682.500000  11.0
                PM           869.000000  10.0
Delta           AM           699.000000  13.0
                PM           897.000000  12.5
Lufthansa       AM           724.666667  10.0
                PM           867.000000  11.0
United          PM           910.000000  12.5

You can choose just a select column or group of columns.

In [224]:
# select only price
flights.pivot_table('price', index=['airline','time_of_day'])

price
airline         time_of_day            
British Airways AM           682.500000
                PM           869.000000
Delta           AM           699.000000
                PM           897.000000
Lufthansa       AM           724.666667
                PM           867.000000
United          PM           910.000000

Or you can compute the average price and time broken down by time of day.

In [226]:
# average price/time broken down by time of day
flights.pivot_table(index=['airline'],columns='time_of_day')

price         time      
time_of_day              AM     PM    AM    PM
airline                                       
British Airways  682.500000  869.0  11.0  10.0
Delta            699.000000  897.0  13.0  12.5
Lufthansa        724.666667  867.0  10.0  11.0
United                  NaN  910.0   NaN  12.5

If you include `margins=True`, it will compute group statistics for all the data within a single tier.

In [228]:
# include margins=True
flights.pivot_table('price', index='airline', columns='time_of_day', margins=True)

time_of_day,AM,PM,All
airline,,,
British Airways,682.500000,869.0,744.666667
Delta,699.000000,897.0,778.200000
Lufthansa,724.666667,867.0,795.833333
United,NaN,910.0,910.000000
All,704.500000,890.4,807.777778


In [232]:
#flights[flights['airline'] == 'British Airways']['price'].mean()
flights[flights['time_of_day'] == 'AM']['price'].mean()
#flights['price'].mean()

704.5

The default function for a pivot table is `mean` although you can change it with the `aggfunc` argument.

In [234]:
# using count
flights.pivot_table(['price'], index=['airline'], columns='time_of_day', margins=True, aggfunc='count')

price          
time_of_day        AM    PM All
airline                        
British Airways   2.0   1.0   3
Delta             3.0   2.0   5
Lufthansa         3.0   3.0   6
United            NaN   4.0   4
All               8.0  10.0  18

A cross-tabulation (or crosstab) is a special case of a pivot table that computes group frequencies.

In [236]:
# using cross tab
pd.crosstab(flights['airline'], flights['time_of_day'], margins=True)

time_of_day,AM,PM,All
airline,,,
British Airways,2,1,3
Delta,3,2,5
Lufthansa,3,3,6
United,0,4,4
All,8,10,18


### Student Practice

Try to perform the following tasks on the `flags` dataset. Then check your answers as I walk through the solutions. 

**Exercise:** Create a pivot table grouping by `religion` in the index and summing the `crescent`, `num_crosses`, and `num_saltires`  columns.

In [258]:
### ENTER CODE HERE ###
#flights.pivot_table('price', index=['airline','time_of_day'])
print(flags.pivot_table('religion', index = ['crescent', 'num_crosses', 'num_saltires'], aggfunc='count'))

##prof solution
print(flags.pivot_table(['num_crosses', 'num_saltires', 'crescent'], index=['religion'], aggfunc='sum'))

                                   religion
crescent num_crosses num_saltires          
0        0           0                  154
                     1                    2
         1           0                   11
                     1                   14
         2           1                    2
1        0           0                   11
          crescent  num_crosses  num_saltires
religion                                     
0                0            2             0
1                0           26            16
2                8            0             0
3                1            1             1
4                1            0             0
5                0            0             1
6                1            0             0
7                0            0             0


**Exercise:** Create a pivot table grouping by `religion` in the index and `zone` in the columns. Use the `sum` function and select the `crescent`, `num_crosses`, and `num_saltires` columns. Add a total for each row and for each column. 

In [260]:
### ENTER CODE HERE ###
#flags.pivot_table('religion', index=['zone'], aggfunc='sum', columns=['crescent', 'num_crosses', 'num_saltires'])

##prof solution
flags.pivot_table(['num_crosses', 'num_saltires', 'crescent'], index=['religion'],
                  columns=['zone'], aggfunc='sum', margins=True)

crescent                    num_crosses                      \
zone            1    2    3    4 All           1    2    3     4 All   
religion                                                               
0             0.0  NaN  0.0  0.0   0         1.0  NaN  0.0   1.0   2   
1             0.0  0.0  0.0  0.0   0         6.0  7.0  4.0   9.0  26   
2             6.0  1.0  NaN  1.0   8         0.0  0.0  NaN   0.0   0   
3             1.0  NaN  NaN  NaN   1         1.0  NaN  NaN   NaN   1   
4             1.0  0.0  NaN  0.0   1         0.0  0.0  NaN   0.0   0   
5             0.0  0.0  NaN  0.0   0         0.0  0.0  NaN   0.0   0   
6             1.0  NaN  NaN  0.0   1         0.0  NaN  NaN   0.0   0   
7             0.0  NaN  NaN  NaN   0         0.0  NaN  NaN   NaN   0   
All           9.0  1.0  0.0  1.0  11         8.0  7.0  4.0  10.0  29   

         num_saltires                     
zone                1    2    3    4 All  
religion                                  
0                 0.0  NaN  0.0  0.0   0  
1                 0.0  5.0  4.0  7.0  16  
2                 0.0  0.0  NaN  0.0   0  
3                 1.0  NaN  NaN  NaN   1  
4                 0.0  0.0  NaN  0.0   0  
5                 0.0  1.0  NaN  0.0   1  
6                 0.0  NaN  NaN  0.0   0  
7                 0.0  NaN  NaN  NaN   0  
All               1.0  6.0  4.0  7.0  18

**Exercise:** Notice that the above output should have a lot of NaNs in the rows. See the [pandas documentation](https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html) to see how to compute the same pivot table but fill all NaNs with a `0`.

In [262]:
### ENTER CODE HERE ###
#flags.pivot_table('religion', index=['zone'], aggfunc='sum', columns=['crescent', 'num_crosses', 'num_saltires'], fill_value=0)

flags.pivot_table(['num_crosses', 'num_saltires', 'crescent'], index=['religion'],
                  columns=['zone'], aggfunc='sum', margins=True, fill_value=0)

crescent              num_crosses               num_saltires        \
zone            1  2  3  4 All           1  2  3   4 All            1  2  3   
religion                                                                      
0               0  0  0  0   0           1  0  0   1   2            0  0  0   
1               0  0  0  0   0           6  7  4   9  26            0  5  4   
2               6  1  0  1   8           0  0  0   0   0            0  0  0   
3               1  0  0  0   1           1  0  0   0   1            1  0  0   
4               1  0  0  0   1           0  0  0   0   0            0  0  0   
5               0  0  0  0   0           0  0  0   0   0            0  1  0   
6               1  0  0  0   1           0  0  0   0   0            0  0  0   
7               0  0  0  0   0           0  0  0   0   0            0  0  0   
All             9  1  0  1  11           8  7  4  10  29            1  6  4   

                 
zone      4 All  
religion         
0         0   0  
1         7  16  
2         0   0  
3         0   1  
4         0   0  
5         0   1  
6         0   0  
7         0   0  
All       7  18

**Exercise:** Using `crosstab`, compute the group frequencies for `landmass` vs `religion`.

In [252]:
### ENTER CODE HERE ###
#pd.crosstab(flights['airline'], flights['time_of_day'], margins=True)
pd.crosstab(flags['landmass'], flags['religion'], margins=True)

religion,0,1,2,3,4,5,6,7,All
landmass,,,,,,,,,
1,9,21,0,0,0,0,1,0,31
2,13,3,0,0,1,0,0,0,17
3,14,13,0,0,0,0,8,0,35
4,2,7,15,0,1,26,0,1,52
5,0,0,20,8,2,0,6,3,39
6,2,16,1,0,0,1,0,0,20
All,40,60,36,8,4,27,15,4,194


In [268]:
iterables = [["bar", "baz", "foo", "qux"], ["one", "two"]]
pd.DataFrame(iterables)

,0,1,2,3
0,bar,baz,foo,qux
1,one,two,None,None


## Building a Machine Learning Model

After thoroughly cleaning and preparing the data, I developed a machine learning model to predict a country's predominant religion based primarily on the characteristics of its flag. Following the creation of the model, I achieved a prediction accuracy of 69% for determining a country's main religion based on the details of its flag. While this accuracy is not exceptionally high, it is understandable given the limited data available for training the model. This result highlights both the potential and the challenges of using symbolic and visual data for cultural predictions.